File: v2/kicad/src/builtins.ts
import type { Pad, PartDefinition, Vec2 } from '../../contracts/src/index.ts';

type Params = Record<string, number | string | boolean>;
export type LocalTrace = { id: string; start: Vec2; end: Vec2; width: number; padId: string; layer: 'front' | 'back' };
export type LocalVia = { id: string; at: Vec2; size: number; drill: number; padId: string };
export type BuiltinGeometry = { pads: Pad[]; courtyard: Vec2[]; traces: LocalTrace[]; vias: LocalVia[] };

const VERSION = '1';
const MIN_HOLE_CLEARANCE = 0.2;
const BOX = (half: number): Vec2[] => [
  { x: -half, y: -half }, { x: half, y: -half },
  { x: half, y: half }, { x: -half, y: half },
];

const numberParam = (params: Params, key: string, fallback: number): number => {
  const value = params[key] ?? fallback;
  if (typeof value !== 'number' || !Number.isFinite(value) || value <= 0) {
    throw new Error(`Invalid ${key} footprint setting`);
  }
  return value;
};

const boolParam = (params: Params, key: string, fallback: boolean): boolean => {
  const value = params[key] ?? fallback;
  if (typeof value !== 'boolean') {
    throw new Error(`Invalid ${key} footprint setting`);
  }
  return value;
};

const pad = (id: string, number: string, x: number, y: number, size: number, drill?: number): Pad => ({
  id, number, at: { x, y }, size: { x: size, y: size }, shape: 'circle', ...(drill ? { drill } : {}),
});

const assertMountClearance = (pads: Pad[]): void => {
  const holes = pads.filter((entry) => entry.plated === false && entry.drill);
  for (const contact of pads.filter((entry) => entry.number)) {
    for (const hole of holes) {
      const dx = Math.abs(contact.at.x - hole.at.x);
      const dy = Math.abs(contact.at.y - hole.at.y);
      const halfX = contact.size.x / 2;
      const halfY = contact.size.y / 2;
      const distance = contact.shape === 'circle'
        ? Math.hypot(dx, dy) - Math.max(halfX, halfY)
        : Math.hypot(Math.max(0, dx - halfX), Math.max(0, dy - halfY));
      if (distance - hole.drill! / 2 < MIN_HOLE_CLEARANCE) {
        throw new Error(`Pad ${contact.number} overlaps mounting hole ${hole.id}`);
      }
    }
  }
};

const switchGeometry = (family: 'mx' | 'choc', hotswap: boolean, params: Params): BuiltinGeometry => {
  const nominalSpacing = family === 'mx' ? (hotswap ? 12.952 : 6.35) : (hotswap ? 11.55 : 5);
  const spacing = numberParam(params, 'padSpacing', nominalSpacing);
  const drill = numberParam(params, 'padDrill', family === 'mx' ? 1.4986 : 1.27);
  const size = numberParam(params, 'padSize', hotswap ? 2.6 : family === 'mx' ? 2.286 : 2.032);
  if (!hotswap && size <= drill) {
    throw new Error('Switch pad size must exceed its drill');
  }
  const reversible = boolParam(params, 'reversible', false);
  const includeCopper = boolParam(params, 'includeTracesVias', false);
  const contacts = family === 'mx'
    ? hotswap
      ? [{ x: 7.11, y: -2.54 }, { x: -5.842, y: -5.08 }]
      : [{ x: -2.54, y: -5.08 }, { x: 3.81, y: -2.54 }]
    : hotswap
      ? [{ x: 3.275, y: -5.95 }, { x: -8.275, y: -3.75 }]
      : [{ x: -5, y: -3.8 }, { x: 0, y: -5.9 }];
  const offset = spacing - nominalSpacing;
  contacts[1].x += Math.sign(contacts[1].x - contacts[0].x) * offset;
  const pads: Pad[] = contacts.map((at, index) => ({
    id: index === 0 ? 'one' : 'two',
    number: String(index + 1),
    at,
    size: { x: size, y: hotswap ? (family === 'mx' ? 2.5 : 2.6) : size },
    shape: hotswap ? 'rect' : 'circle',
    ...(hotswap ? {} : { drill }),
  }));
  const center = family === 'mx' ? 4.1 : 3.4;
  const stabilizer = family === 'mx' ? 1.9 : 1.5;
  const stabilizerX = family === 'mx' ? 5.08 : 5.5;
  pads.push(
    { ...pad('center', '', 0, 0, center, center), plated: false },
    { ...pad('left-stabilizer', '', -stabilizerX, 0, stabilizer, stabilizer), plated: false },
    { ...pad('right-stabilizer', '', stabilizerX, 0, stabilizer, stabilizer), plated: false },
  );
  if (hotswap) {
    const holes = family === 'mx'
      ? [{ x: -2.54, y: -5.08 }, { x: 3.81, y: -2.54 }]
      : [{ x: 0, y: -5.95 }, { x: -5, y: -3.75 }];
    holes.forEach((at, index) => pads.push({ ...pad(`socket-hole-${index}`, '', at.x, at.y, 3, 3), plated: false }));
  }
  assertMountClearance(pads);
  const traces: LocalTrace[] = [];
  const vias: LocalVia[] = [];

  if (reversible && hotswap && includeCopper) {
    const width = numberParam(params, 'traceWidth', 0.25);
    const viaSize = numberParam(params, 'viaSize', 0.8);
    const viaDrill = numberParam(params, 'viaDrill', 0.4);
    if (viaDrill >= viaSize) {
      throw new Error('Via drill must be smaller than via size');
    }
    for (const source of pads.filter((entry) => entry.number)) {
      const at = { x: source.at.x + Math.sign(source.at.x) * 1.8, y: source.at.y };
      traces.push({ id: `route-${source.id}`, start: source.at, end: at, width, padId: source.id, layer: 'front' });
      vias.push({ id: `via-${source.id}`, at, size: viaSize, drill: viaDrill, padId: source.id });
    }
  }

  const courtyard = family === 'choc' && hotswap
    ? [{ x: -10, y: -8 }, { x: 8, y: -8 }, { x: 8, y: 8 }, { x: -10, y: 8 }]
    : BOX(family === 'mx' ? 9 : 8);
  return { pads, courtyard, traces, vias };
};

const diodeGeometry = (): BuiltinGeometry => ({
  pads: [
    { id: 'anode', number: '1', at: { x: -2, y: 0 }, size: { x: 1.2, y: 1.4 }, shape: 'rect' },
    { id: 'cathode', number: '2', at: { x: 2, y: 0 }, size: { x: 1.2, y: 1.4 }, shape: 'rect' },
  ],
  courtyard: [{ x: -3, y: -1.5 }, { x: 3, y: -1.5 }, { x: 3, y: 1.5 }, { x: -3, y: 1.5 }],
  traces: [],
  vias: [],
});

const rgbGeometry = (params: Params): BuiltinGeometry => {
  const spacing = numberParam(params, 'padSpacing', 5.4);
  const width = numberParam(params, 'padSize', 1.1);
  const reversible = boolParam(params, 'reversible', false);
  const includeCopper = boolParam(params, 'includeTracesVias', false);
  const pads = [
    pad('vdd', '1', spacing / 2, 0.7, width),
    pad('dout', '2', spacing / 2, -0.7, width),
    pad('gnd', '3', -spacing / 2, -0.7, width),
    pad('din', '4', -spacing / 2, 0.7, width),
  ];
  const traces: LocalTrace[] = [];
  const vias: LocalVia[] = [];
  if (reversible && includeCopper) {
    const traceWidth = numberParam(params, 'traceWidth', 0.25);
    const viaSize = numberParam(params, 'viaSize', 0.8);
    const viaDrill = numberParam(params, 'viaDrill', 0.4);
    if (viaDrill >= viaSize) {
      throw new Error('Via drill must be smaller than via size');
    }
    for (const source of pads) {
      const at = { x: source.at.x + Math.sign(source.at.x) * 1.8, y: source.at.y };
      traces.push({ id: `route-${source.id}`, start: source.at, end: at, width: traceWidth, padId: source.id, layer: 'front' });
      vias.push({ id: `via-${source.id}`, at, size: viaSize, drill: viaDrill, padId: source.id });
    }
  }
  return { pads, courtyard: BOX(4.5), traces, vias };
};

export function builtinGeometry(source: string, params: Params): BuiltinGeometry | undefined {
  switch (source) {
    case 'builtin:mx-switch': return switchGeometry('mx', false, params);
    case 'builtin:choc-switch': return switchGeometry('choc', false, params);
    case 'builtin:mx-hotswap': return switchGeometry('mx', true, params);
    case 'builtin:choc-hotswap': return switchGeometry('choc', true, params);
    case 'builtin:rgb-led': return rgbGeometry(params);
    case 'builtin:matrix-diode': return diodeGeometry();
    default: return undefined;
  }
}

export function builtinDefinitions(): PartDefinition[] {
  return ([
    ['mx-switch', 'MX switch', 'builtin:mx-switch', 'switch'],
    ['choc-switch', 'Choc switch', 'builtin:choc-switch', 'switch'],
    ['mx-hotswap', 'MX hotswap socket', 'builtin:mx-hotswap', 'connector'],
    ['choc-hotswap', 'Choc hotswap socket', 'builtin:choc-hotswap', 'connector'],
    ['rgb-led', 'RGB LED', 'builtin:rgb-led', 'passive'],
    ['matrix-diode', 'Matrix diode', 'builtin:matrix-diode', 'passive'],
  ] as const).map(([id, name, source, kind]) => {
    const geometry = builtinGeometry(source, {})!;
    return { id, name, kind, ...(kind === 'switch' ? { keycap: { x: 18, y: 18 } } : {}), pads: geometry.pads, courtyard: geometry.courtyard, generator: { source, version: VERSION, parameters: {} } };
  });
}

File: core/src/artifact/builtins.rs
use crate::model::{
    LocalTrace, LocalVia, Pad, PadShape, PartDefinition, PartGenerator, PartKind, Side, Vec2,
};
use serde_json::Value;
use std::collections::BTreeMap;

pub const BUILTIN_VERSION: &str = "1";
const MIN_HOLE_CLEARANCE: f64 = 0.2;

#[derive(Clone, Debug, PartialEq)]
pub struct BuiltinGeometry {
    pub pads: Vec<Pad>,
    pub courtyard: Vec<Vec2>,
    pub traces: Vec<LocalTrace>,
    pub vias: Vec<LocalVia>,
}

fn point(x: f64, y: f64) -> Vec2 {
    Vec2 { x, y }
}

fn box_courtyard(half: f64) -> Vec<Vec2> {
    vec![
        point(-half, -half),
        point(half, -half),
        point(half, half),
        point(-half, half),
    ]
}

fn number_param(params: &BTreeMap<String, Value>, key: &str, fallback: f64) -> Result<f64, String> {
    match params.get(key).map_or(Some(fallback), Value::as_f64) {
        Some(value) if value.is_finite() && value > 0.0 => Ok(value),
        _ => Err(format!("Invalid {key} footprint setting")),
    }
}

fn bool_param(params: &BTreeMap<String, Value>, key: &str, fallback: bool) -> Result<bool, String> {
    match params.get(key) {
        None => Ok(fallback),
        Some(Value::Bool(value)) => Ok(*value),
        _ => Err(format!("Invalid {key} footprint setting")),
    }
}

fn pad(id: &str, number: &str, x: f64, y: f64, size: f64, drill: Option<f64>) -> Pad {
    Pad {
        id: id.into(),
        number: number.into(),
        at: point(x, y),
        size: point(size, size),
        shape: PadShape::Circle,
        drill,
        plated: None,
        side: None,
        rotation: None,
        net_id: None,
    }
}

fn assert_mount_clearance(pads: &[Pad]) -> Result<(), String> {
    let holes = pads
        .iter()
        .filter(|entry| entry.plated == Some(false) && entry.drill.is_some())
        .collect::<Vec<_>>();
    for contact in pads.iter().filter(|entry| !entry.number.is_empty()) {
        for hole in &holes {
            let dx = (contact.at.x - hole.at.x).abs();
            let dy = (contact.at.y - hole.at.y).abs();
            let half_x = contact.size.x / 2.0;
            let half_y = contact.size.y / 2.0;
            let distance = match &contact.shape {
                PadShape::Circle => dx.hypot(dy) - half_x.max(half_y),
                _ => (dx - half_x).max(0.0).hypot((dy - half_y).max(0.0)),
            };
            if distance - hole.drill.expect("mount hole has drill") / 2.0 < MIN_HOLE_CLEARANCE {
                return Err(format!(
                    "Pad {} overlaps mounting hole {}",
                    contact.number, hole.id
                ));
            }
        }
    }
    Ok(())
}

fn switch_geometry(
    family: &str,
    hotswap: bool,
    params: &BTreeMap<String, Value>,
) -> Result<BuiltinGeometry, String> {
    let mx = family == "mx";
    let nominal_spacing = if mx {
        if hotswap { 12.952 } else { 6.35 }
    } else if hotswap {
        11.55
    } else {
        5.0
    };
    let spacing = number_param(params, "padSpacing", nominal_spacing)?;
    let drill = number_param(params, "padDrill", if mx { 1.4986 } else { 1.27 })?;
    let size = number_param(
        params,
        "padSize",
        if hotswap {
            2.6
        } else if mx {
            2.286
        } else {
            2.032
        },
    )?;
    if !hotswap && size <= drill {
        return Err("Switch pad size must exceed its drill".into());
    }
    let reversible = bool_param(params, "reversible", false)?;
    let include_copper = bool_param(params, "includeTracesVias", false)?;
    let mut contacts = if mx {
        if hotswap {
            vec![point(7.11, -2.54), point(-5.842, -5.08)]
        } else {
            vec![point(-2.54, -5.08), point(3.81, -2.54)]
        }
    } else if hotswap {
        vec![point(3.275, -5.95), point(-8.275, -3.75)]
    } else {
        vec![point(-5.0, -3.8), point(0.0, -5.9)]
    };
    let offset = spacing - nominal_spacing;
    contacts[1].x += (contacts[1].x - contacts[0].x).signum() * offset;
    let mut pads = contacts
        .iter()
        .enumerate()
        .map(|(index, at)| Pad {
            id: if index == 0 { "one" } else { "two" }.into(),
            number: (index + 1).to_string(),
            at: *at,
            size: point(
                size,
                if hotswap {
                    if mx { 2.5 } else { 2.6 }
                } else {
                    size
                },
            ),
            shape: if hotswap {
                PadShape::Rect
            } else {
                PadShape::Circle
            },
            drill: if hotswap { None } else { Some(drill) },
            plated: None,
            side: None,
            rotation: None,
            net_id: None,
        })
        .collect::<Vec<_>>();
    let center = if mx { 4.1 } else { 3.4 };
    let stabilizer = if mx { 1.9 } else { 1.5 };
    let stabilizer_x = if mx { 5.08 } else { 5.5 };
    let mut center_pad = pad("center", "", 0.0, 0.0, center, Some(center));
    center_pad.plated = Some(false);
    let mut left = pad(
        "left-stabilizer",
        "",
        -stabilizer_x,
        0.0,
        stabilizer,
        Some(stabilizer),
    );
    left.plated = Some(false);
    let mut right = pad(
        "right-stabilizer",
        "",
        stabilizer_x,
        0.0,
        stabilizer,
        Some(stabilizer),
    );
    right.plated = Some(false);
    pads.extend([center_pad, left, right]);
    if hotswap {
        let holes = if mx {
            vec![point(-2.54, -5.08), point(3.81, -2.54)]
        } else {
            vec![point(0.0, -5.95), point(-5.0, -3.75)]
        };
        for (index, at) in holes.iter().enumerate() {
            let mut hole = pad(
                &format!("socket-hole-{index}"),
                "",
                at.x,
                at.y,
                3.0,
                Some(3.0),
            );
            hole.plated = Some(false);
            pads.push(hole);
        }
    }
    assert_mount_clearance(&pads)?;
    let mut traces = Vec::new();
    let mut vias = Vec::new();
    if reversible && hotswap && include_copper {
        let width = number_param(params, "traceWidth", 0.25)?;
        let via_size = number_param(params, "viaSize", 0.8)?;
        let via_drill = number_param(params, "viaDrill", 0.4)?;
        if via_drill >= via_size {
            return Err("Via drill must be smaller than via size".into());
        }
        for source in pads.iter().filter(|entry| !entry.number.is_empty()) {
            let at = point(source.at.x + source.at.x.signum() * 1.8, source.at.y);
            traces.push(LocalTrace {
                id: format!("route-{}", source.id),
                start: source.at,
                end: at,
                width,
                pad_id: Some(source.id.clone()),
                layer: Side::Front,
            });
            vias.push(LocalVia {
                id: format!("via-{}", source.id),
                at,
                size: via_size,
                drill: via_drill,
                pad_id: Some(source.id.clone()),
            });
        }
    }
    Ok(BuiltinGeometry {
        pads,
        courtyard: if !mx && hotswap {
            vec![
                point(-10.0, -8.0),
                point(8.0, -8.0),
                point(8.0, 8.0),
                point(-10.0, 8.0),
            ]
        } else {
            box_courtyard(if mx { 9.0 } else { 8.0 })
        },
        traces,
        vias,
    })
}

fn diode_geometry() -> BuiltinGeometry {
    BuiltinGeometry {
        pads: vec![
            Pad {
                id: "anode".into(),
                number: "1".into(),
                at: point(-2.0, 0.0),
                size: point(1.2, 1.4),
                shape: PadShape::Rect,
                drill: None,
                plated: None,
                side: None,
                rotation: None,
                net_id: None,
            },
            Pad {
                id: "cathode".into(),
                number: "2".into(),
                at: point(2.0, 0.0),
                size: point(1.2, 1.4),
                shape: PadShape::Rect,
                drill: None,
                plated: None,
                side: None,
                rotation: None,
                net_id: None,
            },
        ],
        courtyard: vec![
            point(-3.0, -1.5),
            point(3.0, -1.5),
            point(3.0, 1.5),
            point(-3.0, 1.5),
        ],
        traces: vec![],
        vias: vec![],
    }
}

fn rgb_geometry(params: &BTreeMap<String, Value>) -> Result<BuiltinGeometry, String> {
    let spacing = number_param(params, "padSpacing", 5.4)?;
    let width = number_param(params, "padSize", 1.1)?;
    let reversible = bool_param(params, "reversible", false)?;
    let include_copper = bool_param(params, "includeTracesVias", false)?;
    let pads = vec![
        pad("vdd", "1", spacing / 2.0, 0.7, width, None),
        pad("dout", "2", spacing / 2.0, -0.7, width, None),
        pad("gnd", "3", -spacing / 2.0, -0.7, width, None),
        pad("din", "4", -spacing / 2.0, 0.7, width, None),
    ];
    let mut traces = Vec::new();
    let mut vias = Vec::new();
    if reversible && include_copper {
        let trace_width = number_param(params, "traceWidth", 0.25)?;
        let via_size = number_param(params, "viaSize", 0.8)?;
        let via_drill = number_param(params, "viaDrill", 0.4)?;
        if via_drill >= via_size {
            return Err("Via drill must be smaller than via size".into());
        }
        for source in &pads {
            let at = point(source.at.x + source.at.x.signum() * 1.8, source.at.y);
            traces.push(LocalTrace {
                id: format!("route-{}", source.id),
                start: source.at,
                end: at,
                width: trace_width,
                pad_id: Some(source.id.clone()),
                layer: Side::Front,
            });
            vias.push(LocalVia {
                id: format!("via-{}", source.id),
                at,
                size: via_size,
                drill: via_drill,
                pad_id: Some(source.id.clone()),
            });
        }
    }
    Ok(BuiltinGeometry {
        pads,
        courtyard: box_courtyard(4.5),
        traces,
        vias,
    })
}

pub fn generate(
    source: &str,
    params: &BTreeMap<String, Value>,
) -> Result<Option<BuiltinGeometry>, String> {
    match source {
        "builtin:mx-switch" => switch_geometry("mx", false, params).map(Some),
        "builtin:choc-switch" => switch_geometry("choc", false, params).map(Some),
        "builtin:mx-hotswap" => switch_geometry("mx", true, params).map(Some),
        "builtin:choc-hotswap" => switch_geometry("choc", true, params).map(Some),
        "builtin:rgb-led" => rgb_geometry(params).map(Some),
        "builtin:matrix-diode" => Ok(Some(diode_geometry())),
        _ => Ok(None),
    }
}

pub fn builtin_definitions() -> Vec<PartDefinition> {
    [
        (
            "mx-switch",
            "MX switch",
            "builtin:mx-switch",
            PartKind::Switch,
        ),
        (
            "choc-switch",
            "Choc switch",
            "builtin:choc-switch",
            PartKind::Switch,
        ),
        (
            "mx-hotswap",
            "MX hotswap socket",
            "builtin:mx-hotswap",
            PartKind::Connector,
        ),
        (
            "choc-hotswap",
            "Choc hotswap socket",
            "builtin:choc-hotswap",
            PartKind::Connector,
        ),
        ("rgb-led", "RGB LED", "builtin:rgb-led", PartKind::Passive),
        (
            "matrix-diode",
            "Matrix diode",
            "builtin:matrix-diode",
            PartKind::Passive,
        ),
    ]
    .into_iter()
    .map(|(id, name, source, kind)| {
        let geometry = generate(source, &BTreeMap::new())
            .expect("built-in defaults validate")
            .expect("known built-in source");
        PartDefinition {
            id: id.into(),
            name: name.into(),
            kind,
            keycap: if id.ends_with("switch") {
                Some(point(18.0, 18.0))
            } else {
                None
            },
            envelope_source: None,
            kicad_source: None,
            terminals: BTreeMap::new(),
            matrix_terminals: None,
            envelope_notice: None,
            courtyard: geometry.courtyard,
            pads: geometry.pads,
            model: None,
            models: None,
            generator: Some(PartGenerator {
                source: source.into(),
                version: BUILTIN_VERSION.into(),
                parameters: BTreeMap::new(),
            }),
        }
    })
    .collect()
}
